lib/ts/recipe/thirdpartyemailpassword/recipe.ts

```diff
@@ -0,0 +1,441 @@
+/* Copyright (c) 2021, VRAI Labs and/or its affiliates. All rights reserved.
+ *
+ * This software is licensed under the Apache License, Version 2.0 (the
+ * "License") as published by the Apache Software Foundation.
+ *
+ * You may not use this file except in compliance with the License. You may
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+import RecipeModule from "../../recipeModule";
+import { NormalisedAppinfo, APIHandled, RecipeListFunction } from "../../types";
+import EmailVerificationRecipe from "../emailverification/recipe";
+import EmailPasswordRecipe from "../emailpassword/recipe";
+import ThirdPartyRecipe from "../thirdparty/recipe";
+import * as express from "express";
+import STError from "./error";
+import { send200Response } from "../../utils";
+import { TypeInput, TypeNormalisedInput, User } from "./types";
+import {
+    validateAndNormaliseUserInput,
+    createNewPaginationToken,
+    combinePaginationTokens,
+    extractPaginationTokens,
+    combinePaginationResults,
+} from "./utils";
+import {
+    SIGN_UP_API,
+    SIGN_IN_API,
+    GENERATE_PASSWORD_RESET_TOKEN_API,
+    PASSWORD_RESET_API,
+    SIGN_OUT_API,
+    SIGNUP_EMAIL_EXISTS_API,
+} from "../emailpassword/constants";
+import { SIGN_IN_UP_API, AUTHORISATION_API } from "../thirdparty/constants";
+import signUpAPI from "../emailpassword/api/signup";
+import signInAPI from "../emailpassword/api/signin";
+import generatePasswordResetTokenAPI from "../emailpassword/api/generatePasswordResetToken";
+import passwordResetAPI from "../emailpassword/api/passwordReset";
+import signOutAPI from "../emailpassword/api/signout";
+import emailExistsAPI from "../emailpassword/api/emailExists";
+import signInUpAPI from "../thirdparty/api/signinup";
+import authorisationUrlAPI from "../thirdparty/api/authorisationUrl";
+import STErrorEmailPassword from "../emailpassword/error";
+import STErrorThirdParty from "../thirdparty/error";
+
+export default class Recipe extends RecipeModule {
+    private static instance: Recipe | undefined = undefined;
+    static RECIPE_ID = "thirdpartyemailpassword";
+
+    config: TypeNormalisedInput;
+
+    emailVerificationRecipe: EmailVerificationRecipe;
+
+    emailPasswordRecipe: EmailPasswordRecipe;
+
+    thirdPartyRecipe: ThirdPartyRecipe | undefined;
+
+    constructor(recipeId: string, appInfo: NormalisedAppinfo, config: TypeInput) {
+        super(recipeId, appInfo);
+        this.config = validateAndNormaliseUserInput(this, appInfo, config);
+        this.emailPasswordRecipe = EmailPasswordRecipe.init({
```

Comment:
Do not call .init for sub recipes. Instead, create a new instance for this and for thirdparty directly.

Reply:
done